src/broadcast.js

```diff
@@ -0,0 +1,230 @@
+   * @param {number} [options.maxSize=100] Defines the max size for the cache messages.
+   */
+  constructor (middleware, options = {}) {
+    assert(middleware);
+    assert(middleware.lookup);
+    assert(middleware.send);
+    assert(middleware.subscribe);
+
+    super();
+
+    const { id = crypto.randomBytes(32), maxAge = 10 * 1000, maxSize = 200 } = options;
+
+    this._id = id;
+    this._lookup = this._buildLookup(middleware.lookup);
+    this._send = (...args) => middleware.send(...args);
+    this._subscribe = onPacket => middleware.subscribe(onPacket);
+
+    this._running = false;
+    this._seenSeqs = new TimeLRUSet({ maxAge, maxSize });
+    this._peers = [];
+    this._codec = new Codec({ verify: true });
+    this._codec.loadFromJSON(schema);
+  }
+
+  /**
+   * Broadcast a flooding message to the peers neighboors.
+   *
+   * @param {Buffer} data
+   * @param {Object} options
+   * @param {Buffer} [options.seqno=crypto.randomBytes(32)]
+   * @returns {Promise}
+   */
+  async publish (data, options = {}) {
+    const { seqno = crypto.randomBytes(32) } = options;
+
+    assert(Buffer.isBuffer(data));
+    assert(Buffer.isBuffer(seqno));
+
+    if (!this._running) {
+      console.warn('Broadcast not running.');
+      return;
+    }
+
+    const packet = { seqno, origin: this._id, data };
+    await this._publish(packet);
+  }
+
+  /**
+   * Initialize the cache and runs the defined subscription.
+   *
+   * @returns {undefined}
+   */
+  run () {
+    if (this._running) return;
+    this._running = true;
+
+    this._subscription = this._subscribe(packetEncoded => this._onPacket(packetEncoded)) || (() => {});
+
+    log('running %h', this._id);
+  }
+
+  /**
+   * Clear the cache and unsubscribe from incoming messages.
+   *
+   * @returns {undefined}
+   */
+  stop () {
+    if (!this._running) return;
+    this._running = false;
+
+    this._subscription();
+    this._seenSeqs.clear();
+
+    log('stop %h', this._id);
+  }
+
+  /**
+   * Build a deferrer lookup.
+   * If we call the lookup several times it would runs once a wait for it.
+   *
+   * @param {Function} lookup
+   * @returns {Function}
+   */
+  _buildLookup (lookup) {
+    let looking = null;
+    return async () => {
+      try {
+        if (!looking) {
+          looking = lookup();
+        }
+        this._peers = await looking;
+        looking = null;
+        log('lookup of %h', this._id, this._peers);
+      } catch (err) {
+        this.emit('lookup-error', err);
+        looking = null;
+      }
+    };
+  }
+
+  /**
+   * Publish and/or Forward a packet message to each peer neighboor.
+   *
+   * @param {Packet} packet
+   * @returns {Promise}
+   */
+  async _publish (packet) {
+    if (!this._running) return;
+
+    try {
+      // Seen it by me.
+      this._seenSeqs.add(msgId(packet.seqno, this._id));
+
+      await this._lookup();
+
+      // I update the package to assign the from prop to me (the current sender).
```

Comment:
Update the package to set the current sender.

Reply:
fixed